File: CoMD/python/cell/comd.py

import simflat
import initatoms
import time
import constants

import argparse
# Translation of CoMD to Python


def advanceVelocity(sim, atoms, dt):
    #for i in range(atoms.nAtoms):
    for iBox in range(sim.boxes.nLocalBoxes):
        for ii in range(sim.boxes.nAtoms[iBox]):
            iOff = sim.boxes.MAXATOMS * iBox + ii
            atoms.p[iOff,0] += dt*atoms.f[iOff,0]
            atoms.p[iOff,1] += dt*atoms.f[iOff,1]
            atoms.p[iOff,2] += dt*atoms.f[iOff,2]

def advancePosition(sim, atoms, dt):
    #for i in range(atoms.nAtoms):
    for iBox in range(sim.boxes.nLocalBoxes):
        for ii in range(sim.boxes.nAtoms[iBox]):
            iOff = sim.boxes.MAXATOMS * iBox + ii
            iSpecies = atoms.species[iOff]
            invMass = 1.0/sim.species[iSpecies].mass
            atoms.r[iOff,0] += dt*atoms.p[iOff,0]*invMass
            atoms.r[iOff,1] += dt*atoms.p[iOff,1]*invMass
            atoms.r[iOff,2] += dt*atoms.p[iOff,2]*invMass


#def redistributeAtoms(sim, atoms):
#    sim.boxes.updateLinkCells(atoms)
#    sim.atomHalo.haloExchange(sim)

def timestep(sim, nSteps, dt):
    for ii in range(nSteps):
        advanceVelocity(sim, sim.atoms, 0.5*dt)
        advancePosition(sim, sim.atoms, dt)
        simflat.redistributeAtoms(sim, sim.atoms)
        sim.pot.computeForce(sim.atoms, sim)
        advanceVelocity(sim, sim.atoms, 0.5*dt)
    #print 'ke,pe = ',initatoms.kineticEnergy(sim)/sim.atoms.nAtoms,sim.ePot/sim.atoms.nAtoms
    sim.eKinetic = initatoms.kineticEnergy(sim)


def initValidate(sim):
    ke = initatoms.kineticEnergy(sim)
    pe = sim.ePot

    print "Initial PE (cohesive energy) : ",pe/sim.atoms.nGlobal
    print "Initial energy : ",(ke+pe)/sim.atoms.nGlobal," atom count : ",sim.atoms.nGlobal

firstCall = True
iStepPrev = -1
def printInfo(sim, iStep, elapsedTime):
    global firstCall, iStepPrev
    if firstCall:
        firstCall = False
        #print "# Loop  Time(fs)   Total Energy  Potential Energy   Kinetic Energy  Temperature  Performance (us/atom)  # Atoms"
        print "#{:>100s}".format("Performance")
        print "#{:>6s} {:>10s} {:>18s} {:>18s} {:>18s} {:>12s} {:^12s} {:^12s}".\
            format("Loop", "Time(fs)", "Total Energy", "Potential Energy", "Kinetic Energy", "Temperature",  "(us/atom)", "# Atoms")

    nEval = iStep - iStepPrev
    iStepPrev = iStep

    simtime = iStep * sim.dt
    n = sim.atoms.nGlobal
    eTotal = (sim.ePot + sim.eKinetic)/n
    eK = sim.eKinetic / n
    eU = sim.ePot / n
    Temp = (sim.eKinetic/n)/(constants.kB_eV * 1.5)
    timePerAtom = 1.0e6 * elapsedTime/(n*nEval)

    print " {:6d} {:10.2f} {:18.12f} {:18.12f} {:18.12f} {:12.4f} {:10.4f} {:12d}".format(iStep, simtime, eTotal, eU, eK, Temp, timePerAtom, n)
    #print iStep, simtime, eTotal, eU, eK, Temp, timePerAtom, n

def printPerformanceResult(sim, elapsedTime):
    n = sim.atoms.nGlobal
    nEval = sim.nSteps
    timePerAtom = 1.0e6 * elapsedTime/(n*nEval)

    print
    print "Average all atom update rate: %10.2f us/atom" % timePerAtom
    print


def parseCommandLine():
    parser = argparse.ArgumentParser()
    parser.add_argument("-x","--nx", type=int, default=20, help="number of unit cells in x")
    parser.add_argument("-y","--ny", type=int, default=20, help="number of unit cells in y")
    parser.add_argument("-z","--nz", type=int, default=20, help="number of unit cells in z")
    parser.add_argument("-N","--nSteps", type=int, default=100, help="total number of time steps")
    parser.add_argument("-n","--printRate", type=int, default=10, help="number of steps between output")
    parser.add_argument("-D","--dt", type=float, default=1, help="time step (in fs)")
    parser.add_argument("-l","--lat", type=float, default=-1, help="lattice parameter (Angstroms)")
    parser.add_argument("-T","--temp", type=float, default=600, help="initial temperature (K)")

    args = parser.parse_args()
    return args


def run_comd():
    cmd = parseCommandLine()
    # init subsystems
    sim = simflat.initSimulation(cmd)

    sim.pot.computeForce(sim.atoms, sim)
    sim.eKinetic = initatoms.kineticEnergy(sim)

    initValidate(sim)

    timestepTime = 0.0
    timestepTimeOneIteration = 0.0

    iStep = 0
    for jStep in range(0, sim.nSteps, sim.printRate):
        printInfo(sim, iStep, timestepTimeOneIteration)

        start = time.clock()
        timestep(sim, sim.printRate, sim.dt)
        end = time.clock()

        timestepTimeOneIteration = end - start
        timestepTime += timestepTimeOneIteration

        iStep += sim.printRate

    printInfo(sim, iStep, timestepTimeOneIteration)

    printPerformanceResult(sim, timestepTime)

    # validate result

if __name__ == '__main__':
    run_comd()

File: CoMD/wolfram/nsquared/comd.m
#!/usr/local/bin/MathematicaScript -script

(* Translation of CoMD to Wolfram Language (Mathematica) *)

Import["ljforce.m"]
Import["initatoms.m"]

nx = 4
ny = 4
nz = 4
temperature = 600 (* in Kelvin *)
dt = 1.0 (* in fs *)
printRate = 10 (* steps between printing output *)
nBlock = 10


(* Setup *)

bounds = {nx, ny, nz} * LJPot`lat
atoms = createFccLattice[nx, ny, nz, LJPot`lat]
p = setTemperature[Length[atoms], LJPot`mass, temperature]

ke = kineticEnergy[p, LJPot`mass];
{F, pe} = LJPot`computeForceSym[atoms, bounds]


(* Formatted output *)

Print[OutputForm[StringForm["Initial PE (cohesive energy) : ``", pe/Length[atoms]]]]
Print[OutputForm[StringForm["Initial energy : `` atom count: ``", (ke+pe)/Length[atoms], Length[atoms]]]]

PadString[s_,nchar_] := StringJoin[PadLeft[Characters[s], nchar, " "]]

PrintHeader := 
(
 Print[OutputForm[PadString["Performance",103]]];
 Print[OutputForm[
    StringJoin[
        "#",
        PadString["Loop",6],
        PadString["Time(fs)",11],
        PadString["Total Energy",19],
        PadString["Potential Energy",19],
        PadString["Kinetic Energy",19],
        PadString["Temperature",13],
        PadString["(us/atom)",13],
        PadString["# Atoms",12]
    ]]]
)

PrintHeader[]

PrintInfo[ke_,pe_,time_,nAtoms_,nStep_,iStep_] := 
(
  eTotal = (ke + pe)/nAtoms;
  timePerAtom = 10^6 * time/(nAtoms*nStep);
  temp = (ke/nAtoms)/(kBeV*1.5);
  (*Print["ke",ke/nAtoms,"pe",pe/nAtoms,"total e",eTotal,"time",OutputForm[time],"us per atom",OutputForm[timePerAtom]]*)
  Print[OutputForm[
   StringJoin[
    ToString[PaddedForm[iStep, 6]],
    ToString[PaddedForm[iStep*dt,{9,2}]],
    ToString[PaddedForm[eTotal,{17,12}]],
    ToString[PaddedForm[pe/nAtoms,{17,12}]],
    ToString[PaddedForm[ke/nAtoms,{17,12}]],
    ToString[PaddedForm[temp,{11,4}]],
    ToString[PaddedForm[timePerAtom,{10,4}]],
    ToString[PaddedForm[nAtoms,11]]
   ]
  ]]
)


(* Advance position and velocity for several timesteps *)

timestep[nSteps_, dt_, atoms_, p_, bounds_, F_] := Module[
{p2=p,atoms2=atoms,F2=F},
(
 Do[
 (
  p2 += 0.5*dt*F2; (* advanceVelocity *)
  atoms2 += dt*p2/LJPot`mass; (* advancePosition *)
  {F2, ePot} = LJPot`computeForceSym[atoms2, bounds];
  p2 += 0.5*dt*F2; (* advanceVelocity *)
 ), {nSteps}];
 {atoms2, p2, F2, ePot})
]

timestepTimeOneIteration = 0.0

(* Main code *)

Do[(
     PrintInfo[ke,pe,timestepTimeOneIteration,Length[atoms],printRate,jStep*printRate];
     {timestepTimeOneIteration, res} =
     Timing[
        ({atoms, p, F, pe} = timestep[printRate, dt, atoms, p, bounds, F];
         ke = kineticEnergy[p, LJPot`mass];
        )
     ]
    ),
    {jStep, nBlock}
  ]

